### PR TITLE
Remove fkey graph visited flags & rework GetConnectedListHelper

### PR DESCRIPTION
**Pull request by onurctirtir**
With this commit, we remove visited flags from ForeignConstraintRelationshipNode
struct since keeping local state in global object is both dangerous and
meaningless.

Also to improve readability, this commit also converts needless recursion to
iterative DFS to avoid passing local hash-map as another parameter to
GetConnectedListHelper function.

See https://github.com/citusdata/citus/pull/4439#discussion_r547855759

### CODE REVIEW

File: src/backend/distributed/utils/foreign_key_relationship.c [hunk 1]
```diff
@@ -264,28 +256,104 @@ SetForeignConstraintRelationshipGraphInvalid()
...
+
+	info.keysize = sizeof(Oid);
+	info.hash = oid_hash;
+	info.hcxt = CurrentMemoryContext;
+
+	/* we don't have value field as it's a set */
+	info.entrysize = info.keysize;
+
+	/* TODO: HASH_FUNCTION needed ? */
+	uint32 hashFlags = (HASH_ELEM | HASH_FUNCTION | HASH_CONTEXT);
```
Comment (onurctirtir) [inline src/backend/distributed/utils/foreign_key_relationship.c:330]: is HASH_FUNCTION needed ?

Comment (marcocitus) [reply src/backend/distributed/utils/foreign_key_relationship.c:330]: We're specifying `info.hash = oid_hash` above, so yes.

These flags pretty much refer to what you specify above.

Comment (onurctirtir) [reply src/backend/distributed/utils/foreign_key_relationship.c:330]: then I can remove this todo, thanks a lot

**Review by marcocitus — APPROVED** [review]
Glad to get this out of the way!

**Pull request merged**